bip-0322.mediawiki

```diff
@@ -160,4 +160,21 @@ This document is licensed under the Creative Commons CC0 1.0 Universal license.
 
 == Test vectors ==
 
-TODO
+=== Message hashing ===
+
+Message hashes are BIP340-tagged hashes of a message, i.e. sha256_tag(m), where tag = <code>BIP0322-signed-message</code>:
+
+* Message = "" (empty string): <code>6e1acf4f6786c3e939571f3e186f45a941208625b78ef293da8fd18327260fff</code>
+* Message = "Hello World": <code>6c5f83704b791f004f0b462ff0b4dd74b4a066facedfe87c12ea3153de5c03d0</code>
+
+=== Message signing ===
+
+Given below parameters:
+
+* private key <code>L3VFeEujGtevx9w18HD1fhRbCH67Az2dpCymeRE1SoPK6XQtaN2k</code>
+* corresponding address <code>bc1q9vza2e8x573nczrlzms0wvx3gsqjx7vavgkx0l</code>
+
+Produce signatures:
+
+* Message = "" (empty string): <code>AkcwRAIgClVQ8S9yX1h8YThlGElD9lOrQbOwbFDjkYb0ebfiq+oCIDHgb/X9WNalNNtqTXb465ufbv9JuLxcJf8qi7DP6yOXASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=</code>
+* Message = "Hello World": <code>AkcwRAIgOi+9FWeR72LaQwZheoopqzQ8apqy2V4J9xtC+WAuMWUCIBxRYlIeSB8B00lzCEwhYIG4oSegnVPWnAmjrBJlNaCQASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=</code>
```

Comment:
I am getting `AkcwRAIgM2gBAQqvZX15ZiysmKmQpDrG83avLIT492QBzLnQIxYCIBaTpOaD20qRlEylyxFSeEA2ba9YOixpX8z46TSDtS40ASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=`

and 

`AkcwRAIgZRfIY3p7/DoVTty6YZbWS71bc5Vct9p9Fia83eRmw2QCICK/ENGfwLtptFluMGs2KsqoNSk89pO7F29zJLUx9a/sASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=`

Reply:
Removing length prefix and only serializing the message string itself I got the same result.

Reply:
Actually, I'm still getting different signatures, even though the hash is the same:

* Message = "" (empty string): <code>AkcwRAIgFuS8y5m0ym9Gj2odoVB5NIL+cPYkeEj8LL1N/6P58X8CIA6jJ9QH2iYKRXVfmhsDzHq1bMS4Adj0nb8DDSdN/SpBASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=</code>
* Message = "Hello World": <code>AkcwRAIgG3PASL/vRTgAqogWT6S8rUOQXNnfRzX6JncmbFlHc1ACIGQdsW+rnVmsQzyAYRQisHKFMigDmKiL7LUw4x17Fw5tASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=</code>

Reply:
What does your `to_spend.vin[0].sigScript` end up as? I'm wondering if the message hash serialization is again doing something unexpected.

Reply:
My end:
```
"" = 0020c90c269c4f8fcbe6880f72a721ddfbf1914268a794cbb21cfafee13770ae19f1
"Hello World" = 0020f0eb03b1a75ac6d9847f55c624a99169b5dccba2a31f5b23bea77ba270de0a7a
```
The spec says

>     vin[0].scriptSig = OP_0 PUSH32[ message_hash ]

which seems to be what I'm doing:
* `00` (OP_0)
* `20` (PUSH32)
* `c90c269c4f8fcbe6880f72a721ddfbf1914268a794cbb21cfafee13770ae19f1` (hash)
*

Reply:
```
"" = 0020c90c269c4f8fcbe6880f72a721ddfbf1914268a794cbb21cfafee13770ae19f1
"Hello World" =  0020f0eb03b1a75ac6d9847f55c624a99169b5dccba2a31f5b23bea77ba270de0a7a
```

maybe something is divering in our to_spend tx creation somehow
```
"" = 00000000010000000000000000000000000000000000000000000000000000000000000000ffffffff220020c90c269c4f8fcbe6880f72a721ddfbf1914268a794cbb21cfafee13770ae19f1000000000100000000000000001600142b05d564e6a7a33c087f16e0f730d1440123799d00000000
"Hello World" = 00000000010000000000000000000000000000000000000000000000000000000000000000ffffffff220020f0eb03b1a75ac6d9847f55c624a99169b5dccba2a31f5b23bea77ba270de0a7a000000000100000000000000001600142b05d564e6a7a33c087f16e0f730d1440123799d00000000

Reply:
For "":
* to_spend = <code>00000000010000000000000000000000000000000000000000000000000000000000000000ffffffff220020c90c269c4f8fcbe6880f72a721ddfbf1914268a794cbb21cfafee13770ae19f1000000000100000000000000001600142b05d564e6a7a33c087f16e0f730d1440123799d00000000</code>
* to_sign = <code>0200000001a7995a543c2866b51ba965e78d01de5db50435cde9d482bf60d8b89ba60a68c5000000000000000000010000000000000000016a00000000</code>

The to_spend looks identical to yours.

Reply:
I realize that `to_sign` is version 2 on my side. It can be either 0 or 2, but this isn't clearly defined anywhere, so unless it's FULL format, you have to guess.

Gut tells me you're correct on this one, i.e. it should use version 0 (disallow locktimes) for SIMPLE format. Changing that, I get

* <code>AkcwRAIgM2gBAQqvZX15ZiysmKmQpDrG83avLIT492QBzLnQIxYCIBaTpOaD20qRlEylyxFSeEA2ba9YOixpX8z46TSDtS40ASECx/EgAxlkQpQ9hYjgGu6EBCPMVPwVIVJqO4XCsMvViHI=</code>

which is the same as yours.